tensorpack/utils/nvml.py

```diff
@@ -0,0 +1,290 @@
+            GeForce GTX 970
+
+        Returns:
+            Name of GPU brand
+        """
+        return _NVML.call(self.hnd, NVML_DEVICE_NAME_BUFFER_SIZE, "nvmlDeviceGetName").value
+
+    def UUID(self):
+        """Return GPU UUID
+
+        Example:
+
+            >>> print(nvidia.Device(0).UUID())
+            GPU-xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+
+        Returns:
+            UUID of GPU
+        """
+        return _NVML.call(self.hnd, NVML_DEVICE_UUID_BUFFER_SIZE, "nvmlDeviceGetUUID").value
+
+    def Memory(self):
+        """Memory information in bytes
+
+        Example:
+
+            >>> print(nvidia.Device(0).Memory())
+            {'total': 4238016512L, 'used': 434831360L, 'free': 3803185152L}
+
+        Returns:
+            total/used/free memory in bytes
+        """
+        class GpuMemoryInfo(Structure):
+            _fields_ = [
+                ('total', c_ulonglong),
+                ('free', c_ulonglong),
+                ('used', c_ulonglong),
+            ]
+
+        c_memory = GpuMemoryInfo()
+        CheckNvmlReturn(_NVML.get_function(
+            "nvmlDeviceGetMemoryInfo")(self.hnd, byref(c_memory)))
+        return {'total': c_memory.total, 'free': c_memory.free, 'used': c_memory.used}
+
+    def Utilization(self):
+        """Percent of time over the past second was utilized.
+
+        Details:
+           Percent of time over the past second during which one or more kernels was executing on the GPU.
+           Percent of time over the past second during which global (device) memory was being read or written
+
+        Example:
+
+            >>> print(nvidia.Device(0).Memory())
+            {'gpu': 4L, 'memory': 6L}
+
+        """
+        class GpuUtilizationInfo(Structure):
+
+            _fields_ = [
+                ('gpu', c_uint),
+                ('memory', c_uint),
+            ]
+
+        c_util = GpuUtilizationInfo()
+        CheckNvmlReturn(_NVML.get_function(
+            "nvmlDeviceGetUtilizationRates")(self.hnd, byref(c_util)))
+        return {'gpu': c_util.gpu, 'memory': c_util.memory}
+
+
+class NvidiaContext(object):
+    """Creates a context to query information
+
+    Example:
+
+        nvidia = NvidiaContext()
+        nvidia.create_context()
+
+        num_gpus = nvidia.NumCudaDevices()
+
+        for device in nvidia.Devices():
+            print(device.Name())
+            print(device.UUID())
+
+            print(device.Memory())
+            print(device.Utilization())
+
+        nvidia.destroy_context()
+    """
+
+    def __init__(self):
+        super(NvidiaContext, self).__init__()
+
+    def create_context(self):
+        """Create a new context
+        """
+        _NVML.load()
+        CheckNvmlReturn(_NVML.get_function("nvmlInit_v2")())
+
+        # Atomically update refcount
+        global nvml_init_count
+        _lib_lock.acquire()
+        nvml_init_count += 1
+        _lib_lock.release()
+
+    def destroy_context(self):
```

Comment:
Make it a with statement?
` with NVMLContext() as ctx: ...`

Reply:
I thought about this as well. But for the callback version of your GPUUtilization call, you cannot use your way, as it closes and open the context for each epoch.